rally-domain/src/main/groovy/yakworks/rally/orgs/repo/ContactRepo.groovy

```diff
@@ -82,21 +87,32 @@ class ContactRepo extends LongIdGormRepo<Contact> {
     @Override
     void doBeforePersistWithData(Contact contact, PersistArgs args) {
         Map data = args.data
+        if (args.bindAction == BindAction.Create) {
+            setupSource(contact, data)
+        }
         // we do primary location and contact here before persist so we persist org only once with contactId it is created
         if(data.location) createOrUpdateLocation(contact, data.location as Map)
     }
 
+    void setupSource(Contact contact, Map data) {
+        if(!contact.id) contact.id = generateId()
+        contact.source = ContactSource.repo.createSource(contact, data)
+    }
+
     /** lookup by num or ContactSource */
     @Override
     Contact lookup(Map data) {
         Contact contact
         if (data == null) data = [:] //if null then make it empty map so it can cycle down and blow error
 
         String sourceId = Maps.value(data, 'sourceId')
-        if(sourceId) {
-            ContactSource source = ContactSource.findWhere(sourceId: sourceId)
-            return source?.contact
-        } else if (data.num) {
+
+        if(data.source == null && sourceId) data.source = [sourceId: sourceId]
```

Comment:
add comment that this a convenience helper that allows sourceId to be passed in at the contact level and not nested into the source object.